### PR TITLE
Market decentralized resolver

### LINKED ISSUE #295: Offer Demand Resolver
### Scope
- [ ] incorporate existing resolver code to yagna repo: https://github.com/stranger80/golem-market-api
- [ ] Resolver is triggered upon Offer or Demand reception:
  - [ ] Subscribe Demand
  - [ ] Subscribe Offer
  - [ ] Broadcast Offer
  - [ ] Retrievie Offers
- [ ] Resover doesn't block execution. It only triggers matching. No awaiting.  
- [ ] new **Proposal**'s are emitted via EventsListeners from Matcher
- [ ] matching do not produce duplicated **Proposal**'s

### Task split
1. resolver with API: `resolve(offers: &Vec<Offer>, demands: &Vec<Demand>) -> Vec<Proposal>`
1. matecher-thread with two APIs:
   - `receive(ofde: OfferDemand)` with just id of an entity, which is stored in DB
   - `Sender<Proposal>` to put resulting Proposals
1. invocation of matcher-thread in four entrypoints specified above and pass `Receiver<Proposal>` to the negotiation engine

related to #232

### PR DESCRIPTION
**Pull request by tworec**
reslover resolves #295 ;)

It glues together changes from #338, #346, #375 and #399.

It also:
- fix query events bugs
- reshapes tests
- introduces DiscoveryBuilder which might be generic solution maybe for Negotiators

### CODE REVIEW

File: core/market/decentralized/src/matcher/resolver.rs [hunk 1]
```diff
@@ -1 +1,122 @@
...
+            process_incoming_subscriptions(subscription_receiver, matcher2move).await
+        });
+
+        Ok(Resolver {
+            matcher: matcher.clone(),
+            subscription_sender,
+        })
+    }
+
+    pub fn incoming_offer(&self, subscription_id: &str) {
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/matcher/resolver.rs:49]: Consider using SubscriptionId type in all our not public interfaces.

File: core/market/decentralized/tests/test_initial_proposal.rs [hunk 2]
```diff
@@ -155,62 +203,43 @@ mod tests {
...
-        };
 
-        // Negative timeout should be treated as immediate checking events and return.
-        let events = tokio::time::timeout(
-            Duration::from_millis(20),
-            market1
-                .requestor_engine
-                .query_events(&demand_id, -5.0, None),
-        )
-        .await??;
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/tests/test_initial_proposal.rs:185]: Why did you removed this case?

File: core/market/decentralized/tests/test_resolver.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        MarketsNetwork,
+    };
+
+    /// Test adds Offer on single node. Resolver should not emit Proposal.
+    #[cfg_attr(not(feature = "market-test-suite"), ignore)]
+    #[actix_rt::test]
+    async fn test_single_not_resolve_offer() -> Result<(), anyhow::Error> {
+        // given
+        let _ = env_logger::builder().try_init();
+        let mut network = MarketsNetwork::new("test_single_resolve_offer")
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/tests/test_resolver.rs:20]: `test_single_resolve_offer` differs from function name `test_single_not_resolve_offer`. For purpose?

File: core/market/decentralized/tests/test_resolver.rs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        let demand = requestor.subscribe_demand(&sample_demand(), &id3).await?;
+
+        // then: It should be resolved on Requestor two times
+        let listener = network.get_event_listeners("Requestor-1");
+        let proposal = timeout200ms(listener.proposal_rx.recv()).await?.unwrap();
+        assert_eq!(proposal.offer, offer1);
+        assert_eq!(proposal.demand, demand);
+        let proposal = timeout200ms(listener.proposal_rx.recv()).await?.unwrap();
+        assert_eq!(proposal.offer, offer2);
+        assert_eq!(proposal.demand, demand);
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/tests/test_resolver.rs:136]: Is ordering deterministic? Or at least - should test depend on this ordering?

Comment (nieznanysprawiciel) [reply core/market/decentralized/tests/test_resolver.rs:136]: But we can leave it

File: core/market/decentralized/tests/utils/mock_net.rs [hunk 5]
```diff
@@ -163,6 +85,78 @@ impl MockNet {
...
+                    "[MockNet] Sending message from [{}], to address [{}].",
+                    &caller,
+                    &local_addr
+                );
+                Ok(local_bus::send(&local_addr, &caller, &data).await?)
+            }
+        });
+    }
+}
+
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/tests/utils/mock_net.rs:159]: Is it just copy, or should I read it?

File: core/market/decentralized/src/db/dao/events.rs [hunk 6]
```diff
@@ -12,16 +12,11 @@ use crate::db::DbResult;
...
-    EventsRemovalError {
-        num_removed: usize,
-        to_remove: usize,
-    },
     #[error("Subscription [{0}] not found. Could be unsubscribed.")]
     SubscriptionNotFound(SubscriptionId),
     #[error("Subscription [{0}] expired.")]
     SubscriptionExpired(SubscriptionId),
-    #[error(transparent)]
+    #[error("Failed to get events from database. Error: {0}.")]
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/db/dao/events.rs:19]: so do I, that's why I changed your code ;)

File: core/market/decentralized/src/db/dao/events.rs [hunk 7]
```diff
@@ -70,31 +65,17 @@ impl<'c> EventsDao<'c> {
...
-                        num_removed,
-                        to_remove: events.len(),
-                    };
-                    log::error!("{}", error);
-                    return Err(error);
+            // Remove returned events from queue.
+            if !events.is_empty() {
+                let ids = events.iter().map(|event| event.id).collect::<Vec<_>>();
+                if !ids.is_empty() {
+                    diesel::delete(dsl::market_event.filter(dsl::id.eq_any(ids))).execute(conn)?;
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/db/dao/events.rs:72]: Diesel statements map quite literally to SQL

File: core/market/decentralized/src/db/dao/events.rs [hunk 8]
```diff
@@ -70,31 +65,17 @@ impl<'c> EventsDao<'c> {
...
-                    let error = TakeEventsError::EventsRemovalError {
-                        num_removed,
-                        to_remove: events.len(),
-                    };
-                    log::error!("{}", error);
-                    return Err(error);
+            // Remove returned events from queue.
+            if !events.is_empty() {
+                let ids = events.iter().map(|event| event.id).collect::<Vec<_>>();
+                if !ids.is_empty() {
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/db/dao/events.rs:71]: will remove this check

File: core/market/decentralized/src/lib.rs [hunk 9]
```diff
@@ -23,14 +25,3 @@ pub mod migrations {
...
-/// These exports are expected to be used only in tests.
-pub mod testing {
-    pub use super::db::dao::{DemandDao, OfferDao};
-    pub use super::db::models::{Demand, Offer, SubscriptionId, SubscriptionParseError};
-    pub use super::matcher::{DemandError, MatcherError, OfferError};
-    pub use super::matcher::{DraftProposal, SubscriptionStore};
-    pub use super::negotiation::QueryEventsError;
-    pub use super::negotiation::{ProviderBroker, RequestorBroker};
-    pub use super::protocol::negotiation;
-}
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/lib.rs:36]: because I've moved `mock_offer.rs` from `tests` to be testing sub-module and wanted to hide it in subdir.

File: core/market/decentralized/src/market.rs [hunk 10]
```diff
@@ -55,7 +56,7 @@ impl MarketService {
 
         let (matcher, listeners) = Matcher::new(db)?;
         let provider_engine = ProviderBroker::new(db.clone())?;
-        let requestor_engine = RequestorBroker::new(db.clone(), listeners.proposal_receiver)?;
+        let requestor_engine = RequestorBroker::new(db.clone(), listeners.proposal_rx)?;
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/market.rs:59]: `tx, rx` is a de-facto naming standard for the `sender, receiver`. See
```
grep --include '*.rs' -rn 'let[^=]*rx\>' agent core exe-unit service-bus test-utils utils
```
which gives 34 results across all yagna repo, comparing to only 6 from
```
grep --include '*.rs' -rn 'let[^=]*receiver\>' agent core exe-unit service-bus test-utils utils
```
3 of which are from the decentralized market still

File: core/market/decentralized/src/matcher.rs [hunk 11]
```diff
@@ -148,26 +146,27 @@ impl Matcher {
...
+    resolver
+        .store
         .store_offer(msg.offer)
         .await
         .map(|propagate| match propagate {
-            true => Propagate::Yes,
+            true => {
+                resolver.receive(subscription).unwrap();
+                Propagate::Yes
+            }
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher.rs:169]: will fix

File: core/market/decentralized/src/matcher/error.rs [hunk 12]
```diff
@@ -56,6 +64,16 @@ pub enum MatcherInitError {
     DatabaseError(#[from] DbError),
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum ResolverError {
+    #[error("Failed resolve matching relation for {0:?} and {1:?}.")]
+    MatchingError(Offer, Demand),
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/error.rs:71]: yes, this uses debug `:?`

File: core/market/decentralized/src/matcher/error.rs [hunk 13]
```diff
@@ -56,6 +64,16 @@ pub enum MatcherInitError {
...
+    #[error("Failed to process incoming subscription {0:?}")]
+    SendError(#[from] tokio::sync::mpsc::error::SendError<Subscription>),
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/error.rs]: Good point. There will be no subscription id which caused the problem. I Will enrich this err msg it.

File: core/market/decentralized/src/matcher/resolver.rs [hunk 14]
```diff
@@ -1 +1,136 @@
...
+        Ok(self.subscription_tx.send(subscription.into())?)
+    }
+
+    // TODO: it is mocked; emits dummy RawProposal upon every Subscription
+    async fn process_incoming_subscriptions(
+        &self,
+        mut subscription_rx: UnboundedReceiver<Subscription>,
+    ) {
+        while let Some(subscription) = subscription_rx.recv().await {
+            log::debug!("processing incoming subscription {:?}", subscription);
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/resolver.rs]: this was on purpose. Subscription does not support display. And it is an enum with name of subscription id: Offer or Demand

File: core/market/decentralized/src/matcher/resolver.rs [hunk 15]
```diff
@@ -1 +1,136 @@
...
+            match subscription {
+                Subscription::Offer(id) => {
+                    log::info!("TODO: resolve new Offer: {:?}", id);
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/resolver.rs]: yup, it was leftover.

File: core/market/decentralized/src/matcher/store.rs [hunk 16]
```diff
@@ -58,6 +58,37 @@ impl SubscriptionStore {
...
+    pub async fn get_offers_before(&self, demand: &Demand) -> Result<Vec<Offer>, OfferError> {
+        let now = Utc::now().naive_utc();
+        Ok(self
+            .db
+            .as_dao::<OfferDao>()
+            .get_offers_before(demand.insertion_ts.unwrap(), now)
+            .await
+            .map_err(|e| OfferError::GetMany(e))?)
+    }
+
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:91]: Good point.

The intent was to show the purpose or the context in which this is usable. `insertion_ts` is designed for that single purpose to be compared against other Demand[s], not between Offers themselves. 

But now I see I can go even further and enclose getting single Demand and all Offers before it within single offer store fn. Since it is the only usecase for it and will illustrate the intent clearly. Furthermore, it will ensure that `unwrap` is safe.

File: core/market/decentralized/src/matcher/store.rs [hunk 17]
```diff
@@ -58,6 +58,37 @@ impl SubscriptionStore {
...
     }
 
+    pub async fn get_offers(&self, id: Option<Identity>) -> Result<Vec<ClientOffer>, OfferError> {
+        Ok(self
+            .db
+            .as_dao::<OfferDao>()
+            .get_offers(match id {
+                Some(ident) => Some(ident.identity),
+                _ => None,
+            })
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:68]: nice :) thx

[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:68]: This is @jiivan code. As I've merged to his PR, but still I will apply this simplification.

File: core/market/decentralized/src/matcher/resolver.rs [hunk 18]
```diff
@@ -1 +1,136 @@
...
+            // TODO: should we stop processing events / panic ?
+            log::error!(
+                "Failed to emit proposal [offer_id:{}, demand_id:{}]: {}",
+                offer_id,
+                demand_id,
+                e
+            )
+        };
+    }
+}
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/resolver.rs]: Thanks for this comment. This enabled me to find a bug, bc current code will break the whole matching when one pair (offer, demand) matching causes an error, and this should not be broken but to be continued for all remaining pairs.

File: core/market/decentralized/src/matcher/store.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:91]: but for the time being, I'm simplifying the API as you've suggested

File: core/market/decentralized/src/db/dao/events.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/db/dao/events.rs:19]: :facepalm:

File: core/market/decentralized/src/market.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/market.rs:59]: I know and it's terrible. At least don't change my names

File: core/market/decentralized/src/matcher/error.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher/error.rs:71]: I don't think, it's good idea. I wouldn't print such an error later. It pollutes whole log and we can't see details whatsoever

[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher/error.rs:71]: Hmm if you want to have them printed somewhere, you could make log::trace in place where this error happens, but return in error only ids

File: core/market/decentralized/src/negotiation/errors.rs [hunk 19]
```diff
@@ -31,13 +29,3 @@ pub enum QueryEventsError {
...
-
-impl From<TakeEventsError> for QueryEventsError {
-    fn from(e: TakeEventsError) -> Self {
-        match e {
-            TakeEventsError::SubscriptionExpired(id) => QueryEventsError::SubscriptionExpired(id),
-            TakeEventsError::SubscriptionNotFound(id) => QueryEventsError::Unsubscribed(id),
-            _ => QueryEventsError::FailedGetEvents(e),
-        }
-    }
-}
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/negotiation/errors.rs:43]: I've changed it because it has flaw. If we've got `TakeEventsError::SubscriptionNotFound` we cannot translate it into `QueryEventsError::Unsubscribed` because it is not only possibility. Subscription might never existed. This change is not only for code brevity, but also for correctness

File: core/market/decentralized/src/protocol/discovery.rs [hunk 20]
```diff
@@ -242,7 +231,7 @@ pub enum Propagate {
     No(Reason),
 }
 
-#[derive(Clone, Serialize, Deserialize)]
+#[derive(Clone, Serialize, Deserialize, Debug)]
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/protocol/discovery.rs:234]: probably some debugging. It is a leftover. Removing

File: core/market/decentralized/src/protocol/discovery/builder.rs [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+    pub fn add_data_handler<M: CallbackMessage, T: Clone + Send + Sync + 'static>(
+        mut self,
+        mut f: impl DataCallbackHandler<M, T>,
+    ) -> Self {
+        let boxed = self
+            .data
+            .get(&TypeId::of::<T>())
+            .expect("data handler needs data");
+
+        // let data = *(boxed as &Box<dyn Any + 'static>).downcast::<&T>().unwrap();
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/protocol/discovery/builder.rs:30]: nope, removing

File: core/market/decentralized/src/protocol/discovery/builder.rs [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    pub fn add_data_handler<M: CallbackMessage, T: Clone + Send + Sync + 'static>(
+        mut self,
+        mut f: impl DataCallbackHandler<M, T>,
+    ) -> Self {
+        let boxed = self
+            .data
+            .get(&TypeId::of::<T>())
+            .expect("data handler needs data");
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/protocol/discovery/builder.rs:28]: Because this only can happen when the Market developer (one of us) is misusing the builder (forgotten to specify data). It is not user-facing error. it should fail fast during initial testing or demo

File: core/market/decentralized/src/rest_api/error.rs [hunk 23]
```diff
@@ -69,8 +79,11 @@ impl ResponseError for QueryEventsError {
     fn error_response(&self) -> HttpResponse {
         let msg = ErrorMessage::new(self.to_string());
         match self {
-            QueryEventsError::Unsubscribed(_) => HttpResponse::NotFound().json(msg),
-            QueryEventsError::SubscriptionExpired(_) => HttpResponse::NotFound().json(msg),
+            QueryEventsError::Unsubscribed(_)
+            | QueryEventsError::TakeEventsError(TakeEventsError::SubscriptionNotFound(_))
+            | QueryEventsError::TakeEventsError(TakeEventsError::SubscriptionExpired(_)) => {
+                HttpResponse::NotFound().json(msg)
+            }
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/rest_api/error.rs:86]: I've chosen not to redefine  those two `TakeEventsError` values within  `QueryEventsError`... for the code brevity

File: core/market/decentralized/src/negotiation/errors.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/negotiation/errors.rs:43]: Than you should rather change QueryEventsError to `SubscriptionNotFound`, because we can never return `QueryEventsError::Unsubscribed`

[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/negotiation/errors.rs:43]: I don't like this pattern of embedding errors in errors in every place possible. It makes sense in some places, but in some doesn't. QueryEventsError should give very specific errors without need to dispatch nesting.

File: core/market/decentralized/src/protocol/discovery/builder.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/protocol/discovery/builder.rs:28]: And the same happens if you return error from initialization

File: core/market/decentralized/src/rest_api/error.rs [hunk 24]
```diff
@@ -40,6 +42,14 @@ impl ResponseError for MatcherError {
...
 impl ResponseError for NegotiationError {}
 
+impl ResponseError for ResolverError {
+    fn error_response(&self) -> HttpResponse {
+        match self {
+            _ => HttpResponse::InternalServerError().json(ErrorMessage::new(self.to_string())),
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/rest_api/error.rs:52]: Good catch! Thx

File: core/market/decentralized/tests/test_initial_proposal.rs [hunk 25]
```diff
@@ -2,21 +2,40 @@ mod utils;
...
-    use crate::utils::mock_offer::{example_demand, example_offer};
-    use crate::utils::MarketsNetwork;
+    use crate::utils::{MarketServiceExt, MarketsNetwork};
 
-    use std::sync::Arc;
-    use std::time::Duration;
+    /// No events available for not existent subscription.
+    #[cfg_attr(not(feature = "market-test-suite"), ignore)]
+    #[actix_rt::test]
+    async fn test_query_events_non_existent_subscription() -> Result<(), anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/tests/test_initial_proposal.rs:17]: I like naming and isolation of edge cases

File: core/market/decentralized/src/negotiation/errors.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/negotiation/errors.rs:43]: > Than you should rather change QueryEventsError to `SubscriptionNotFound`, because we can never return `QueryEventsError::Unsubscribed`
yes, we can and it is illustrated two times in test


> I don't like this pattern of embedding errors in errors in every place possible. It makes sense in some places, but in some doesn't. QueryEventsError should give very specific errors without need to dispatch nesting.

Here it seems to me it is one of the case when embedding works very well, because otherwise you need to repeat all patterns from `TakeEventsError` in `QueryEventsError`, isn't it?

File: core/market/decentralized/src/protocol/discovery/builder.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/protocol/discovery/builder.rs:28]: Nope, in intialisation you are also handling GSB errors... it might be user-facing

File: core/market/decentralized/tests/test_resolver.rs (hunk 3 above)
Comment (tworec) [reply core/market/decentralized/tests/test_resolver.rs:20]: thx

File: core/market/decentralized/tests/test_resolver.rs (hunk 4 above)
Comment (tworec) [reply core/market/decentralized/tests/test_resolver.rs:136]: again good catch. It caused an error and I've fixed it

File: core/market/decentralized/tests/utils/mock_net.rs (hunk 5 above)
Comment (tworec) [reply core/market/decentralized/tests/utils/mock_net.rs:159]: copy

File: core/market/decentralized/tests/test_initial_proposal.rs (hunk 2 above)
Comment (tworec) [reply core/market/decentralized/tests/test_initial_proposal.rs:185]: I've removed because we need to wait a bit for resolver to process subscriptions. Haven't considered to obviously wait before calling it :) restoring

File: core/market/decentralized/src/market.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/market.rs:59]: it is also in code examples within Tokio eg.
https://docs.rs/tokio/0.2.21/tokio/sync/mpsc/struct.UnboundedReceiver.html

I'm reverting within your code if it is hurting you so much

File: core/market/decentralized/src/matcher/error.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/error.rs:71]: ok

File: core/market/decentralized/src/matcher/error.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/error.rs]: removed this enum option altogether

File: core/market/decentralized/src/negotiation/errors.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/negotiation/errors.rs:43]: Repeating isn't a problem for me. Ugly api is a problem.

And what are patterns on which caller wants to match against here?
_NotFound_, _Unsubscribed_, _Expired_ are fundamental errors and caller can react somehow to them.
Other errors aren't so important, because caller could threat them as "some internal error" and he can't do anything more with this. He can just print message and wait for human to read it ;)

In this code the most important errors are hidden behind strange enum `TakeEventsError`. `TakeEventsError` tells him nothing, because he already knows that querying events failed. So he must dive deeper, to know what happened.

Embedding makes sense, when we deal with errors from internal layers. We want to wrap them with higher level errors, to add more meaningful message. When there are multiple layers, we have full path of errors that caused problem.
In such case wrapping adds information. Here we take information.

File: core/market/decentralized/src/protocol/discovery/builder.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/protocol/discovery/builder.rs:28]: Ok, if you insists ;)

File: core/market/decentralized/src/negotiation/errors.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/negotiation/errors.rs:43]: I think we could leave it, but maybe we should make some meeting to establish common error handling practices, that would satisfy us all

File: core/market/decentralized/src/matcher.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher.rs]: This flaw was introduced originally in #290 :-P
https://github.com/golemfactory/yagna/pull/290/files#diff-c373b03f716ab716b09699faa1c50599R227-R255

**Comment by tworec** [comment]
@nieznanysprawiciel pls review again

File: core/market/decentralized/src/db/dao/offer.rs [hunk 26]
```diff
@@ -58,6 +92,25 @@ impl<'c> OfferDao<'c> {
         .await
     }
 
+    pub async fn checked_insert(
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/db/dao/offer.rs:95]: All inserts should be checked. No need to highlight this in name

File: core/market/decentralized/src/matcher.rs [hunk 27]
```diff
@@ -148,26 +145,27 @@ impl Matcher {
...
     // not only Offers from other nodes.
 
-    store
-        .store_offer(msg.offer)
+    let subscription = Subscription::from(&msg.offer);
+    resolver
+        .store
+        .save_offer(msg.offer)
         .await
         .map(|propagate| match propagate {
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/matcher.rs:164]: This should not be named `propagate`, but probably `inserted` or something like this. Save_offer doesn't return info, if we should propagate anything

File: core/market/decentralized/src/matcher/error.rs [hunk 28]
```diff
@@ -56,6 +62,14 @@ pub enum MatcherInitError {
     DatabaseError(#[from] DbError),
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum ResolverError {
+    #[error(transparent)]
+    OfferError(#[from] OfferError),
```
Comment (nieznanysprawiciel) [inline core/market/decentralized/src/matcher/error.rs:68]: OfferError has ResolverError and ResolverError has OfferError

File: core/market/decentralized/src/db/dao/offer.rs (hunk 26 above)
Comment (tworec) [reply core/market/decentralized/src/db/dao/offer.rs:95]: initial should not I believe...

Comment (nieznanysprawiciel) [reply core/market/decentralized/src/db/dao/offer.rs:95]: Probably we should have only one function. Caller decides if this failing check is an error or valid situation.
But doesn't matter. Let's merge this quickly

File: core/market/decentralized/src/matcher.rs (hunk 27 above)
Comment (tworec) [reply core/market/decentralized/src/matcher.rs:164]: true

File: core/market/decentralized/src/matcher/error.rs (hunk 28 above)
Comment (tworec) [reply core/market/decentralized/src/matcher/error.rs:68]: This shouldn't be possible. `MatcherError` has `ResolveError` which has `OfferError`

Comment (tworec) [reply core/market/decentralized/src/matcher/error.rs:68]: I've tested and such setup results in compile error:
```
error[E0391]: cycle detected when processing `matcher::error::OfferError`
```

File: core/market/decentralized/src/db/dao/offer.rs (hunk 26 above)
Comment (tworec) [reply core/market/decentralized/src/db/dao/offer.rs]: Ok, now there is a single insert function which returns resulting Offer with `insertion_ts`.

**Comment by tworec** [comment]
A few improvements:
- cleaned unused errors
- divided huge `OfferError` into specialized errors eq. `QueryOfferError`
- moved unsubscribe offer logic to `SubscriptionStore` which is capable of deciding when to remove Offer and when to just mark it as unsubscribed.

@nieznanysprawiciel pls re-review
